src/core/Akka.Streams/Dsl/StreamRefs.cs

```diff
@@ -0,0 +1,729 @@
+    /// The receiving end of these messages MUST fail the stream if it observes gaps in the sequence,
+    /// as these messages will not be re-delivered.
+    /// 
+    /// Sequence numbers start from `0`.
+    /// </summary>
+    internal sealed class SequencedOnNext : IStreamRefsProtocol, IDeadLetterSuppression
+    {
+        public long SeqNr { get; }
+        public object Payload { get; }
+
+        public SequencedOnNext(long seqNr, object payload)
+        {
+            SeqNr = seqNr;
+            Payload = payload ?? throw ReactiveStreamsCompliance.ElementMustNotBeNullException;
+        }
+    }
+
+    /// <summary>
+    /// Initial message sent to remote side to establish partnership between origin and remote stream refs.
+    /// </summary>
+    internal sealed class OnSubscribeHandshake : IStreamRefsProtocol, IDeadLetterSuppression
+    {
+        public OnSubscribeHandshake(IActorRef targetRef)
+        {
+            TargetRef = targetRef;
+        }
+
+        public IActorRef TargetRef { get; }
+    }
+
+    /// <summary>
+    /// Sent to a the receiver side of a stream ref, once the sending side of the SinkRef gets signalled a Failure.
+    /// </summary>
+    internal sealed class RemoteStreamFailure : IStreamRefsProtocol
+    {
+        public RemoteStreamFailure(string message)
+        {
+            Message = message;
+        }
+
+        public string Message { get; }
+    }
+
+    /// <summary>
+    /// Sent to a the receiver side of a stream ref, once the sending side of the SinkRef gets signalled a completion.
+    /// </summary>
+    internal sealed class RemoteStreamCompleted : IStreamRefsProtocol
+    {
+        public RemoteStreamCompleted(long seqNr)
+        {
+            SeqNr = seqNr;
+        }
+
+        public long SeqNr { get; }
+    }
+
+    /// <summary>
+    /// INTERNAL API: Cumulative demand, equivalent to sequence numbering all events in a stream.
+    /// 
+    /// This message may be re-delivered.
+    /// </summary>
+    internal sealed class CumulativeDemand : IStreamRefsProtocol, IDeadLetterSuppression
+    {
+        public CumulativeDemand(long seqNr)
+        {
+            if (seqNr < 0) throw ReactiveStreamsCompliance.NumberOfElementsInRequestMustBePositiveException;
+            SeqNr = seqNr;
+        }
+
+        public long SeqNr { get; }
+    }
+
+    #endregion
+
+    #region extension
+
+    internal sealed class StreamRefsMaster : IExtension
+    {
+        public static StreamRefsMaster Get(ActorSystem system) =>
+            system.WithExtension<StreamRefsMaster, StreamRefsMasterProvider>();
+
+        private readonly EnumerableActorName sourceRefStageNames = new EnumerableActorNameImpl("SourceRef", new AtomicCounterLong(0L));
+        private readonly EnumerableActorName sinkRefStageNames = new EnumerableActorNameImpl("SinkRef", new AtomicCounterLong(0L)); 
+        
+        public StreamRefsMaster(ExtendedActorSystem system)
```

Comment:
Can the parameter be removed since it’s not stored ?

Reply:
I just want to keep the convention for extensions here.